### PR TITLE
EQ: Add equalizer control tool eqctl

### PR DESCRIPTION
**Pull request by singalsu**
This patch adds a simple control tool to setup during runtime (idle)
the IIR or FIR equalizer with new coefficients configuration. It is
useful in tuning of the equalizers. The tool also has capability to read
the EQ configuration from DSP.

Run "eqctl -h" to see usage. The directory SOFT/tune/eq contains tool
to create EQ setup data.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
I just pushed an update. There was a typo in eqctl/Makefile.am that somehow didn't break build. Now it's corrected.

File: eqctl/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,6 @@
+LDADD = -lasound
+
+bin_PROGRAMS = eqctl
```
Comment (lgirdwood) [inline eqctl/Makefile.am]: can you change the bin name to "sof-eqctl". This will avoid naming collisions.

File: eqctl/eqctl.c [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <alsa/asoundlib.h>
+
+#define SOF_CTRL_CMD_BINARY 3 /* TODO: From uapi ipc */
+
+static void usage(char *name)
+{
+	printf("Usage %s <option(s)>\n", name);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: fprintf(stdout, is better

File: eqctl/eqctl.c [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	int n_max = smax / sizeof(unsigned int);
+	int separator;
+
+	fh = fopen(setup, "r");
+	while (fscanf(fh, "%u", &data[n]) != EOF && n < n_max) {
+		if (n > 0)
+			printf(",");
+		printf("%u", data[n]);
+		separator = fgetc(fh);
+		//printf("<%c>", (char)separator);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: debug to be deleted.

File: eqctl/eqctl.c [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+	/* Open the device, mixer control and get read/write/type properties.
+	 */
+	ret = snd_ctl_open(&ctl, dev, mode);
+	if (ret) {
+		fprintf(stderr, "Error: Could not open device %s.\n", dev);
+		exit(EXIT_FAILURE);
+	}
+
+	snd_ctl_elem_info_alloca(&info);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: pls comment what this block does.

File: eqctl/eqctl.c [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+	snd_ctl_elem_id_alloca(&id);
+	snd_ctl_elem_value_alloca(&value);
+	ret = snd_ctl_ascii_elem_id_parse(id, cname);
+	if (ret) {
+		fprintf(stderr, "Error: Can't find %s.\n", cname);
+		exit(EXIT_FAILURE);
+	}
+
+	snd_ctl_elem_info_set_id(info, id);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: ditto

File: eqctl/eqctl.c [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	ret = snd_ctl_elem_info(ctl, info);
+	if (ret) {
+		fprintf(stderr, "Error: Could not get elem info.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	ctrl_size = snd_ctl_elem_info_get_count(info);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: ditto

File: eqctl/eqctl.c [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	read = snd_ctl_elem_info_is_tlv_readable(info);
+	write = snd_ctl_elem_info_is_tlv_writable(info);
+	type = snd_ctl_elem_info_get_type(info);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: better to report if it's not readable and writable in separate messages as that can help diagnose permissions issues.

File: eqctl/eqctl.c [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+	/* Next allocate buffer for tlv write/read. The buffer needs a two
+	 * words header with tag (SOF_CTRL_CMD_BINARY) and size in bytes.
+	 */
+	buffer_size = ctrl_size + 2 * sizeof(unsigned int);
+	user_data = calloc(1, buffer_size);
+	user_data[0] = SOF_CTRL_CMD_BINARY;
+	if (set) {
+		fprintf(stderr, "Applying configuration \"%s\" ", setup);
+		fprintf(stderr, "into device %s control %s.\n", dev, cname);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: stdout

File: eqctl/eqctl.c [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+		if (ret) {
+			fprintf(stderr, "Error: failed TLV write.\n");
+			free(user_data);
+			exit(EXIT_FAILURE);
+		}
+		fprintf(stderr, "Success.\n");
+
+	} else {
+		fprintf(stderr, "Retrieving configuration for ");
+		fprintf(stderr, "device %s control %s.\n", dev, cname);
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: stdout

File: eqctl/eqctl.c [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+		}
+		fprintf(stderr, "Success.\n");
+		config = (uint32_t *)(user_data + 2);
+		n = config[0] / sizeof(uint32_t);
+		for (i = 0; i < n; i++) {
+			if (i == n - 1)
+				printf("%u\n", config[i]);
+			else
+				printf("%u,", config[i]);
+		}
```
Comment (lgirdwood) [inline eqctl/eqctl.c]: what does this block do, pls comment

File: eqctl/eqctl.c (hunk 2 above)
Comment (singalsu) [reply eqctl/eqctl.c]: ok, there goes may line length saving trick :^)

File: eqctl/eqctl.c (hunk 3 above)
Comment (singalsu) [reply eqctl/eqctl.c]: yep

File: eqctl/eqctl.c (hunk 4 above)
Comment (singalsu) [reply eqctl/eqctl.c]: OK, I'm adding comments. Please check from next version that I got it right.

File: eqctl/eqctl.c (hunk 7 above)
Comment (singalsu) [reply eqctl/eqctl.c]: ok!

File: eqctl/eqctl.c (hunk 8 above)
Comment (singalsu) [reply eqctl/eqctl.c]: OK, I'll change. This was a trick to get coefficients back to a correct syntax file without other information.

Comment (singalsu) [reply eqctl/eqctl.c]: But better to add later file output with e.g. -o if need for convenience. I'm not also totally sure about the ASCII file format for numbers data but use this for now.

File: sof-eqctl/README [hunk 11]
```diff
@@ -0,0 +1,77 @@
+How to use sof-eqctl tool
```
Comment (lgirdwood) [inline sof-eqctl/README]: This should be in the sof docs repo.

File: sof-eqctl/sof-eqctl.c [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+	FILE *fh;
+	int n = 0;
+	int n_max = smax / sizeof(unsigned int);
+	int separator;
+
+	fh = fopen(setup, "r");
+	while (fscanf(fh, "%u", &data[n]) != EOF && n < n_max) {
+		if (n > 0)
+			printf(",");
+		printf("%u", data[n]);
```
Comment (lgirdwood) [inline sof-eqctl/sof-eqctl.c]: fprintf(stdout

File: sof-eqctl/sof-eqctl.c [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+		fprintf(stderr, "Error: Can't find %s.\n", cname);
+		exit(EXIT_FAILURE);
+	}
+
+        /* Get handle info from id. */
+	snd_ctl_elem_info_set_id(info, id);
+	ret = snd_ctl_elem_info(ctl, info);
+	if (ret) {
+		fprintf(stderr, "Error: Could not get elem info.\n");
+		exit(EXIT_FAILURE);
```
Comment (lgirdwood) [inline sof-eqctl/sof-eqctl.c]: exit(ret) ? and in other cases

File: Makefile.am [hunk 14]
```diff
@@ -1,4 +1,4 @@
-SUBDIRS = rmbox topology
+SUBDIRS = rmbox topology sof-eqctl
```
[the remark this answers is not in the document]
Comment (singalsu) [reply Makefile.am]: Yes, there was issues I fought for a while to make "-" to work :^)

**Comment by singalsu** [comment]
I have just pushed fixed for your review feedback. I'll make also a PR into sof-docs for most of the previous README file content.

**Comment by lgirdwood** [comment]
@singalsu IThere was a merge conflict too, but I seem to have resolved it with the github merge too ?

**Comment by singalsu** [comment]
@lgirdwood Looks like I need to rebase and fix conflict. It's not merged.

**Comment by singalsu** [comment]
Travis will likely fail with missing /usr/local/include/sof/uapi/logging.h.

**Comment by xiulipan** [comment]
@singalsu 
Travis CI is expect to be fail since logger need SOF `make install` first.
Will try to fix this issue https://github.com/thesofproject/soft/issues/85 ASAP.

**Pull request merged**